### PR TITLE
Implement DialogView for Uap

### PR DESCRIPTION
**Pull request by andrechi1**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
New feature. Implement ModalView for Uap

I haven't been able to prove it well, because the Playground.Uwp project doesn't start.

### :thinking: Checklist before submitting

- [ ] All projects build
- [ ] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [ ] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contributing/mvvmcross-docs-style-guide))
- [ ] Rebased onto current develop

### CODE REVIEW

**Comment by martijn00** [comment]
@nickrandolph can you have a look at why the playground doesn't start?

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 1]
```diff
@@ -23,6 +23,9 @@ public class MvxWindowsViewPresenter
     {
         protected readonly IMvxWindowsFrame _rootFrame;
 
+        private MvxWindowsContentDialog mDialogView;
```
Comment (nmilcoff) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Can we avoid by any chance holding a strong reference to the dialog view here?

Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: I don't know how to do it. You have access to the ViewModel, but not to the View.

**Comment by andrechi1** [comment]
To run the Playground, throw this exception. I don't now if this error is reported.

MvvmCross.Exceptions.MvxException
  HResult=0x80131500
  Message=Problem navigating to ViewModel RootViewModel
  Source=MvvmCross
  StackTrace:
   at MvvmCross.ViewModels.MvxAppStart`1.<NavigateToFirstViewModel>d__1.MoveNext() in D:\Apl\Xamarin\MvvmCross\MvvmCross\MvvmCross\ViewModels\MvxAppStart.cs:line 83
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MvvmCross.ViewModels.MvxAppStart.<StartAsync>d__5.MoveNext() in D:\Apl\Xamarin\MvvmCross\MvvmCross\MvvmCross\ViewModels\MvxAppStart.cs:line 43
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MvvmCross.ViewModels.MvxAppStart.Start(Object hint) in D:\Apl\Xamarin\MvvmCross\MvvmCross\MvvmCross\ViewModels\MvxAppStart.cs:line 28
   at MvvmCross.Platforms.Uap.Views.MvxApplication.RunAppStart(IActivatedEventArgs activationArgs) in D:\Apl\Xamarin\MvvmCross\MvvmCross\MvvmCross\Platforms\Uap\Views\MvxApplication.cs:line 72
   at MvvmCross.Platforms.Uap.Views.MvxApplication.OnLaunched(LaunchActivatedEventArgs activationArgs) in D:\Apl\Xamarin\MvvmCross\MvvmCross\MvvmCross\Platforms\Uap\Views\MvxApplication.cs:line 46

Inner Exception 1:
MvxException: Failed to construct and initialize ViewModel for type Playground.Core.ViewModels.RootViewModel from locator MvxDefaultViewModelLocator - check InnerException for more information

Inner Exception 2:
MvxException: Problem running viewModel lifecycle of type RootViewModel

Inner Exception 3:
NullReferenceException: Object reference not set to an instance of an object.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 1 above)
Comment (Cheesebaron) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Doesn't follow the naming conventions we use.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 2]
```diff
@@ -240,5 +251,40 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
                 return Task.FromResult(false);
             }
         }
+
+        protected virtual async Task<bool> ShowModal(Type viewType, MvxBasePresentationAttribute attribute, MvxViewModelRequest request)
+        {
+            try
+            {
+                var modalAttribute = (MvxModalViewPresentationAttribute)attribute;
+                var instanceReques = (MvxViewModelInstanceRequest)request;
```
Comment (Cheesebaron) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Move this cast closer to usage. Also typo.

File: MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        {
+        }
+
+        private IMvxViewModel _viewModel;
+
+        public IMvxViewModel ViewModel
+        {
+            get
+            {
+                return _viewModel;
```
Comment (Cheesebaron) [inline MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs]: Can you please use Expression body

File: MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    public class MvxWindowsContentDialog<TViewModel>
+        : MvxWindowsContentDialog
+        , IMvxView<TViewModel> where TViewModel : class, IMvxViewModel
+    {
+        public new TViewModel ViewModel
+        {
+            get { return (TViewModel)base.ViewModel; }
```
Comment (Cheesebaron) [inline MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs]: Expression body

File: MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        ~MvxWindowsContentDialog()
+        {
+            Dispose(false);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
```
Comment (Cheesebaron) [inline MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs]: Remove this if, not used at all.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 6]
```diff
@@ -240,5 +250,39 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
                 return Task.FromResult(false);
             }
         }
+
+        protected virtual async Task<bool> ShowModal(Type viewType, MvxBasePresentationAttribute attribute, MvxViewModelRequest request)
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Should you pass in MvxModalViewPresentationAttribute as type?

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 7]
```diff
@@ -240,5 +250,39 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
...
+        {
+            try
+            {
+                if (Activator.CreateInstance(viewType) is MvxWindowsContentDialog modalView)
+                {
+                    var instanceReques = (MvxViewModelInstanceRequest)request;
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Typo

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 8]
```diff
@@ -23,6 +23,8 @@ public class MvxWindowsViewPresenter
     {
         protected readonly IMvxWindowsFrame _rootFrame;
 
+        private MvxWindowsContentDialog _windowsContentDialog;
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: I don't think it is good to keep an instance reference like this. We probably need something to find the view back in the stack.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 9]
```diff
@@ -240,5 +250,39 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
...
+            {
+                MvxLog.Instance.Trace("Error seen during navigation request to {0} - error {1}", request.ViewModelType.Name,
+                    exception.ToLongString());
+                return false;
+            }
+        }
+
+        protected virtual Task<bool> CloseModal(IMvxViewModel viewModel, MvxBasePresentationAttribute attribute)
+        {
+            _windowsContentDialog?.Hide();
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Is Hide correct, or do we really need to remove it as well?

File: MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+// The .NET Foundation licenses this file to you under the MS-PL license.
+// See the LICENSE file in the project root for more information.
+
+using MvvmCross.ViewModels;
+using MvvmCross.Views;
+using Windows.UI.Xaml.Controls;
+
+namespace MvvmCross.Platforms.Uap.Views
+{
+    public class MvxWindowsContentDialog
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Views/MvxWindowsContentDialog.cs]: The view is missing lifecycle methods.

File: MvvmCross/Platforms/Uap/Presenters/Attributes/MvxModalViewPresentationAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MS-PL license.
+// See the LICENSE file in the project root for more information.
+
+using MvvmCross.Presenters.Attributes;
+using Windows.UI.Xaml.Controls;
+
+namespace MvvmCross.Platforms.Uap.Presenters.Attributes
+{
+    public sealed class MvxModalViewPresentationAttribute : MvxBasePresentationAttribute
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/Attributes/MvxModalViewPresentationAttribute.cs]: Don't make it sealed

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 9 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: I think it's the right one. PrimaryButtonCommand can be invoked, but it's rare.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 8 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Made.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 12]
```diff
@@ -55,6 +58,14 @@ public override void RegisterAttributeTypes()
                    ShowAction = (view, attribute, request) => ShowRegionView(view, (MvxRegionPresentationAttribute)attribute, request),
                    CloseAction = (viewModel, attribute) => CloseRegionView(viewModel, (MvxRegionPresentationAttribute)attribute)
                });
+
+            AttributeTypesToActionsDictionary.Add(
+               typeof(MvxModalViewPresentationAttribute),
+               new MvxPresentationAttributeAction
+               {
+                   ShowAction = (view, attribute, request) => ShowModal(view, (MvxModalViewPresentationAttribute)attribute, (MvxViewModelInstanceRequest)request),
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Should not cast to MvxViewModelInstanceRequest here, but in the method

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 13]
```diff
@@ -240,5 +251,36 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
                 return Task.FromResult(false);
             }
         }
+
+        protected virtual async Task<bool> ShowModal(Type viewType, MvxModalViewPresentationAttribute attribute, MvxViewModelInstanceRequest request)
+        {
+            try
+            {
+                if (Activator.CreateInstance(viewType) is MvxWindowsContentDialog modalView)
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Isn't there a view creator for Windows? @nickrandolph

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 14]
```diff
@@ -240,5 +251,36 @@ protected virtual Task<bool> ShowPage(Type viewType, MvxBasePresentationAttribut
...
+            {
+                MvxLog.Instance.Trace("Error seen during navigation request to {0} - error {1}", request.ViewModelType.Name,
+                    exception.ToLongString());
+                return false;
+            }
+        }
+
+        protected virtual Task<bool> CloseModal(IMvxViewModel viewModel, MvxBasePresentationAttribute attribute)
+        {
+            var popups = VisualTreeHelper.GetOpenPopups(Window.Current).FirstOrDefault(p => p.Child is ContentDialog);
```
Comment (martijn00) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Check for `MvxWindowsContentDialog` to get a closer match. If possible also check the viewmodel type or something to make sure we don't close the wrong one when multiple modals are open.

Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: You cannot open two ContentDialog at a time.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 12 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: I do not understand why attribute yes and request no. What criterion do you have?

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 13 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: ContentDialog is created with the new command. But if you know another way to do it, propose it.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 12 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: +1

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs (hunk 13 above)
Comment (andrechi1) [reply MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: The ContentDialog not is a page.  I can create a method to create the ContentDialog. I can't think of anything else.

File: MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs [hunk 15]
```diff
@@ -256,14 +256,15 @@ protected virtual async Task<bool> ShowModal(Type viewType, MvxModalViewPresenta
         {
             try
             {
-                if (Activator.CreateInstance(viewType) is MvxWindowsContentDialog modalView)
+                IMvxWindowsContentDialog contentDialog;
+                if ((contentDialog = CreateContentDialog(attribute, viewType)) != null)
```
Comment (Cheesebaron) [inline MvvmCross/Platforms/Uap/Presenters/MvxWindowsViewPresenter.cs]: Please split assignment and null check

**Comment by andrechi1** [comment]
I changed the modal name to dialog. A ContentDialog is not a modal window. And it can confuse.

**Comment by andrechi1** [comment]
This is a sample for can use x:Bind in XAML.

**Comment by nickrandolph** [comment]
Apologies been unavailable to contribute. A quick glance - I think in the close method we should be confirming that the open dialogue is for the viewmodel being closed

**Comment by andrechi1** [comment]
If it leaves you calmer. Here's the modification. But you can only open one ContentDialog per window. 

I'm trying to be able to open Standalone windows. And if you can open two ContentDialog. But they are placed in different visual trees. When you closed this change I send the window modal. I need help because you have to modify MvxWindowsMainThreadDispatcher.

**Comment by martijn00** [comment]
@andrechi1 @nickrandolph @Cheesebaron i've done a rebase and implemented the changes to get this merged in.

**Pull request merged**